File: bindings/r/R/OnlineLoess.R
#' LOESS Online Smoothing
#'
#' @description
#' Create a stateful LOESS model for real-time online data.
#'
#' @srrstats {G2.0} Input validation for fraction, window_capacity, min_points.
#' @srrstats {G1.6} Sliding window for incremental updates.
#'
#' @inheritParams Loess
#' @param window_capacity Maximum number of points kept in the sliding window.
#' @param min_points Minimum number of points required before smoothing begins.
#' @param update_mode Window update strategy: \code{"full"} (default) re-smooths
#'   all window points after each addition, \code{"incremental"} updates only the
#'   newest point.
#' @param confidence_intervals Confidence level for confidence intervals
#'   (e.g., 0.95). \code{NULL} (default) disables confidence intervals.
#' @param prediction_intervals Confidence level for prediction intervals
#'   (e.g., 0.95). \code{NULL} (default) disables prediction intervals.
#'
#' @return An OnlineLoess object.
#' @examples
#' model <- OnlineLoess(fraction = 0.2, window_capacity = 20)
#' x <- 1:50
#' y <- sin(x * 0.1) + rnorm(50, 0, 0.1)
#' result <- model$add_points(x, y)
#' plot(x, y)
#' lines(x, result$y, col = "red")
#' @export
OnlineLoess <- function(
    fraction = 0.2,
    window_capacity = 100L,
    min_points = 2L,
    iterations = 3L,
    weight_function = "tricube",
    robustness_method = "bisquare",
    scaling_method = "mad",
    boundary_policy = "extend",
    update_mode = "full",
    auto_converge = NULL,
    return_robustness_weights = FALSE,
    return_diagnostics = FALSE,
    return_residuals = FALSE,
    zero_weight_fallback = "use_local_mean",
    parallel = FALSE,
    degree = "linear",
    dimensions = 1L,
    distance_metric = "normalized",
    surface_mode = "interpolation",
    return_se = FALSE,
    confidence_intervals = NULL,
    prediction_intervals = NULL,
    weighted_metric_weights = NULL,
    cell = NULL,
    interpolation_vertices = NULL,
    boundary_degree_fallback = NULL
) {
    validate_params(
        fraction = fraction, window_capacity = window_capacity,
        min_points = min_points
    )
    handle <- do.call(ROnlineLoess$new, env_args(online_params))

    structure(
        list(
            handle = handle,
            add_points = function(x, y) {
                args <- validate_common_args(x, y, fraction, iterations)
                handle$add_points(args$x, args$y)
            },
            params = list(
                fraction = fraction,
                window_capacity = window_capacity,
                min_points = min_points,
                iterations = iterations,
                parallel = parallel
            )
        ),
        class = "OnlineLoess"
    )
}

File: bindings/java/src/main/java/fastloess/OnlineLoess.java
package fastloess;

import java.util.Optional;

/**
 * An online LOESS model that updates incrementally as points arrive.
 */
public final class OnlineLoess implements AutoCloseable {

    private long handle;

    /**
     * Creates a new online model from the given options.
     *
     * @param options the model configuration
     */
    public OnlineLoess(OnlineOptions options) {
        Options c = options.common;
        this.handle = NativeBridge.onlineNew(
                c.fraction,
                c.iterations,
                c.weightFunction,
                c.robustnessMethod,
                c.scalingMethod,
                c.boundaryPolicy,
                c.returnRobustnessWeights,
                c.zeroWeightFallback,
                c.autoConverge,
                c.parallel,
                options.windowCapacity,
                options.minPoints,
                options.updateMode,
                c.degree,
                c.dimensions,
                c.distanceMetric,
                c.surfaceMode,
                c.cell,
                c.interpolationVertices,
                NativeBridge.boolSentinel(c.boundaryDegreeFallback),
                c.weightedMetricWeights);
    }

    /**
     * Adds a point to the model, returning a smoothed output once at least
     * {@code minPoints} have been seen, or {@link Optional#empty()} otherwise.
     *
     * @param x the x value
     * @param y the y value
     * @return the smoothed output, or {@link Optional#empty()} if not enough
     * points have been seen yet
     */
    public Optional<PointResult> addPoint(double x, double y) {
        checkOpen();
        NativeOnlineOutput o = NativeBridge.onlineAddPoint(handle, x, y);
        return o.hasValue ? Optional.of(PointResult.fromNative(o)) : Optional.empty();
    }

    private void checkOpen() {
        if (handle == 0) {
            throw new IllegalStateException("OnlineLoess has already been closed");
        }
    }

    @Override
    public void close() {
        if (handle != 0) {
            NativeBridge.onlineFree(handle);
            handle = 0;
        }
    }
}
